src/release/breaking-changes/android-14-nonlinear-text-scaling.md

```diff
@@ -0,0 +1,328 @@
+|-------------------------------------------------------------------------------|--------------------------------------------------------|
+| `SelectableText({ double? TextScaleFactor = 1.0 })` constructor argument      | 'textScaleFactor' is deprecated and shouldn't be used. |
+| `SelectableText.rich({ double? TextScaleFactor = 1.0 })` constructor argument | 'textScaleFactor' is deprecated and shouldn't be used. |
+| `SelectableText.textScaleFactor` getter                                       | 'textScaleFactor' is deprecated and shouldn't be used. |
+
+## Migration guide
+
+Widgets provided by the Flutter framework are already migrated. Migration is 
+needed only if you're using any of the deprecated symbols listed above. 
+
+However, the new Android 14 nonlinear scaling feature changes the UI of existing 
+apps, even if your app does not directly use any of the deprecated APIs. If your 
+app targets Android 14 devices, it is **strongly recommended** to test your app 
+on Android 14 devices at different scaling levels, to make sure the UI is usable 
+and functional, no matter whether the codebase needs migration or not. Please 
+refer to the [Ensure Android 14 compatibility][] guide for steps to make sure 
+your app is Android 14 compatible.
+
+As an example, the following widget might not look great on Android 14 at 200% 
+scaling:
+
+```dart 
+// It's not good for tooltips to be cutoff vertically, or to take too much space 
+// on screen as it blocks contents.
+MyTooltipBox( 
+  size: tooltipBoxSize * textScaleFactor,
+  child: RichText(..., style: TextStyle(fontSize: 20)),
+)
+```
+
+With nonlinear scaling at 200%, the actual font size used by the `RichText` 
+widget is 21 (instead of 40), and the `MyTooltipBox` widget would take 
+significantly more screen space than needed, potentially making the tooltip 
+annoying. On the other hand, at a smaller font size, the tooltip's text could be 
+cutoff due to insufficient vertical space.
+
+Adopting nonlinear text scaling might not be an easy task. If you're unsure about 
+how to migrate certian `textScaleFactor` usages, or when it is going to take an
+extended period of time (for example, when the UI has to be redesigned), 
+consider replacing `textScaleFactor` with `TextScaler.textScaleFactor` first. The 
+`TextScaler.textScaleFactor` getter was added for backward compatibility during 
+the migration process and is expected to have a longer lifetime than other 
+`textScaleFactor` deprecations. 
+
+### Migrating your APIs that expose `textScaleFactor`
+
+Before:
+```dart 
+abstract class _MyCustomPaintDelegate { 
+  void paint(PaintingContext context, Offset offset, double textScaleFactor) { 
+  }
+}
+```
+
+After:
+```dart 
+abstract class _MyCustomPaintDelegate { 
+  void paint(PaintingContext context, Offset offset, TextScaler textScaler) { 
+    // For now, implementers can still access the `textScaleFactor`.
+    // scalar via textScaler.textScaleFactor. 
```

Comment:
Let's focus the guide on what the right way to migrate is, e.g. how should textScaler be used correctly?

Reply:
Ideally, this guide shouldn't mention `textScaler.textScaleFactor` at all since that is not the right way to migrate your code long-term. If necessary, we can at the very end mention it for completeness, but it I wouldn't focus the guide this much around it. It's not a long-term solution and asking people to migrate to something that is already deprecated is somewhat frustrating.

Reply:
Should I add a section for apps that don't target Android 14 and tell them to add an extension?

Reply:
Is that realistic for an app to not target Android 14? That'll have to happen sooner or later, no? I don't think I would spend too much time on how to do the thing we don't want developers to do...

Reply:
that's true for desktop-only apps or web-only apps I think?

Reply:
lol, ok. The long-term goal is that TextScaler is the only text scaling mechanism in Flutter, though. And I don't think we should (pretend to) support anything else. The migration guide should show people how to migrate to TextScaler, which is the API we want people to use and the only one we will actually support in the future.

Reply:
I'll save that for the migration guide for nonlinear text scaling